brainspell/article_helpers.py

```diff
@@ -67,25 +69,91 @@ def update_vote_in_struct(struct, tag_name, username, direction, label_name):
+    if target.count == 0:
+        Votes.insert(username=username,
+                     name=topic,
+                     article_id=pmid,
+                     vote=direction,
+                     type=True).execute()
+        if direction:  # Increase number of agreements
+            Tags.update(agree=Tags.agree + 1)\
+                .where((Tags.tag_name == topic) &
+                       (Tags.article_id == pmid) &
+                       (Tags.experiment_id is None)
+                       ).execute()
+        else:
+            Tags.update(disagree=Tags.disagree + 1).\
+                where((Tags.tag_name == topic)
+                      & (Tags.article_id == pmid)
+                      & (Tags.experiment_id is None)
+                      ).execute()
+    elif target.count > 0:
+        target = next(target)
+        # Remove the vote (constitutes a double select)
+        if target.vote == direction:
+            Votes.delete().where((Votes.username == username) &
+                                 (Votes.name == topic) &
+                                 (Votes.article_id == pmid)).execute()
+            # Remove the vote from the agree/disagree fields of Articles Mesh
+            # tags
+            if direction:  # If vote is an agreement, decrement the agreement field
+                Tags.update(
+                    agree=Tags.agree -
+                    1).where(
+                    (Tags.tag_name == topic) & (
+                        Tags.article_id == pmid) & (
+                        Tags.experiment_id is None)).execute()
+            else:
+                Tags.update(
+                    disagree=Tags.disagree -
+                    1).where(
+                    (Tags.tag_name == topic) & (
+                        Tags.article_id == pmid) & (
+                        Tags.experiment_id is None)).execute()
+        else:
+            # Toggle the vote
+            Votes.update(
+                vote=direction).where(
+                (Votes.username == username) & (
+                    Votes.name == topic) & (
+                    Votes.article_id == pmid)).execute()
+            if direction:  # This indicates that we went from downvoting to upvoting
+                Tags.update(agree=Tags.agree + 1, disagree=Tags.disagree - 1)\
+                    .where((Tags.tag_name == topic)
+                           & (Tags.article_id == pmid)
+                           & (Tags.experiment_id is None)
+                           ).execute()
+            else:
+                Tags.update(agree=Tags.agree - 1, disagree=Tags.disagree + 1)\
+                    .where((Tags.tag_name == topic)
+                           & (Tags.article_id == pmid)
+                           & (Tags.experiment_id is None)
+                           ).execute()
 
 
 def vote_stereotaxic_space(pmid, space, username):
+    # TODO: This method doesn't really make sense anymore now that space is
```

Comment:
Should we delete the method?

Reply:
We might want to make a variant of this so the front-end passes back the experimentID